api/lib.go

```diff
@@ -76,11 +77,12 @@ func Instantiate(cache Cache, code_id []byte, params []byte, msg []byte, store K
 	db := buildDB(store)
 	a := buildAPI(api)
 	q := buildQuerier(querier)
-	var gasUsed u64
+	var gasUsed = u64(0)
 	errmsg := C.Buffer{}
 	res, err := C.instantiate(cache.ptr, id, p, m, db, a, q, u64(gasLimit), &gasUsed, &errmsg)
-	if err != nil {
-		return nil, 0, errorWithMessage(err, errmsg)
+	if err != nil && err.(syscall.Errno) != C.ErrnoValue_Success {
```

Comment:
Why the `&&`... This means if `err == nil && err.(syscall.Errno) == C.ErrnoValue_Success` then we return `nil` for the error and parse the result?

Maybe this can never happen (in which case the check is unneeded) and if it can happen, acting like a success is probably not the desired behavior.

Reply:
I think you mixed up `==` and `!=`?
You are correct that because `ErrnoValue_Success` is just `0`, then it means that returning it from Rust will make `err == nil`, but my rationale was "what if success wasn't signalled with 0?". It's not likely to ever be the case, so i don't have a problem with removing the downcast and comparison to `ErrnoValue_Success`.

Reply:
Not mixed up, I was looking for a counter-example.

Anyway, yes `errno != 0` => `err != nil`, so let's just keep the if guard here the same as before.

Reply:
alright, cool :)